### PR TITLE
auto create ldap role(&group) for auto registered user

### LINKED ISSUE #5129: Assign (and create) roles to users based on their LDAP group
AD/LDAP can contain quite some information about users. One useful attribute that may be available is one describing the users membership to a 'unit' or 'group'.

If the Auth providers are allowed to return Group/Roles names to which the user will be automatically assigned, it can reduce admin manual work.

In addition, one might think to also auto create a data library based on this information - but this is even more specific and out of the scope for this issue.

### PR DESCRIPTION
**Pull request by scholtalbers**
fix #5129

### CODE REVIEW

**Comment by bernt-matthias** [comment]
Cool. Would be really useful for me. Originally I intended to use this https://github.com/bernt-matthias/galaxy-group-sync, but your PR seems to solve this much cleaner. 

Two questions: 
- what happens if LDAP roles and groups change? 
- what do I need to do in order to test this? (I have a dev-instance for testing which is configured to use LDAP)

**Comment by scholtalbers** [comment]
@bernt-matthias this only works for newly registered users and does not sync after. I.e. roles will get outdated if not manually maintained. For this your `sync` methods seems fit. Otherwise this could be done upon every login, similar to what I do here: https://github.com/galaxyproject/galaxy/pull/5138/commits/58c7bdaca0f0f28d88ea76ebbfda5d679fdf8a91#diff-4ee48e7674940f5ea4a6486f7d680e7cR584

One thing to note as well, this creates roles and groups. But only assigns the Role to the user (guess we could just assign both).

For you to test, see the changes made in the [auth_conf.xml](https://github.com/galaxyproject/galaxy/pull/5136/commits/aa3983bf9152a78bf2b577a1a4f3e51ecc387c64#diff-c48ef3414328ae988f799895ff6ab002R28). You only need to add two lines.

**Comment by jmchilton** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

File: config/auth_conf.xml.sample [hunk 1]
```diff
@@ -88,12 +104,14 @@
...
             <auto-register-username>{sAMAccountName}</auto-register-username>
             <auto-register-email>{mail}</auto-register-email>
+            <auto-register-roles>gidNumber</auto-register-roles>
 -->
             <!-- For OpenLDAP: -->
 <!--        <bind-user>{dn}</bind-user>
             <bind-password>{password}</bind-password>
             <auto-register-username>{uid}</auto-register-username>
             <auto-register-email>{mail}</auto-register-email>
+            <auto-register-roles>gid</auto-register-roles>
```
Comment (natefoo) [inline config/auth_conf.xml.sample:114]: It's not readily apparent what this value does - the LDAP response's `gid` value is used as the name for the Galaxy group and/or role? Why is this not templated like other values?

File: lib/galaxy/auth/__init__.py [hunk 2]
```diff
@@ -102,11 +108,21 @@ def check_auto_registration(self, trans, login, password):
...
+                        options.get('auto-assign-roles-to-groups-only', False))
+
+                    if len(auth_results) == 4:
+                        auth_return["attributes"] = auth_results[3]
+                    return auth_return
                 elif auth_result is None:
                     log.debug("Email: %s, Username %s, stopping due to failed non-continue" % (auto_email, auto_username))
                     break  # end authentication (skip rest)
-        return (False, '', '')
+        return auth_return
```
Comment (natefoo) [inline lib/galaxy/auth/__init__.py:125]: I worry that the growing complexity of this return value structure will cause some extra logic in callers (although I guess there's only one caller now?). Plus the `auth_results` from the provider could be troublesome for other plugins or future plugin development. Maybe a class that can be shared for both purposes should be used instead?

File: lib/galaxy/auth/providers/ldap_ad.py [hunk 3]
```diff
@@ -92,6 +92,20 @@ def authenticate(self, email, username, password, options):
                 log.debug('LDAP authenticate: email must be used to login, cannot be None')
                 return (failure_mode, '', '')
 
+        auto_create_roles = string_as_bool(options.get('auto-create-roles', False))
+        auto_create_groups = string_as_bool(options.get('auto-create-groups', False))
+        auto_create_roles_or_groups = auto_create_roles or auto_create_groups
+        auto_assign_roles_to_groups_only = string_as_bool(options.get('auto-assign-roles-to-groups-only', False))
+        if auto_assign_roles_to_groups_only and not (auto_create_roles and auto_create_groups):
+            raise ConfigurationError("If 'auto-assign-roles-to-groups-only' is True, auto-create-roles and "
+                                     "auto-create-groups have to be True as well.")
```
Comment (natefoo) [inline lib/galaxy/auth/providers/ldap_ad.py:101]: These aren't LDAP-specific are they? If not, they should be handled in `AuthManager` I believe.

File: lib/galaxy/security/__init__.py [hunk 4]
```diff
@@ -776,6 +776,34 @@ def get_private_user_role(self, user, auto_create=False):
                 return None
         return role
 
+    def get_admin_role(self, name):
```
Comment (natefoo) [inline lib/galaxy/security/__init__.py:779]: Maybe `def get_role(self, name, type=self.model.Role.types.ADMIN):` and similar below for `create_admin_role()`, so these are more usable in the general case.

**Review by natefoo — COMMENTED** [review]
Thanks @scholtalbers! I think this a useful feature and it looks pretty good. I have a few comments inline.

**Comment by scholtalbers** [comment]
Will close in favor of #5238

**Pull request closed without merging**